src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -233,13 +247,70 @@ private void cancel() {
 
     }
 
+    private class Callback2 extends BodyExecutionCallback {
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            if (killer != null) {
+                killer.cancel(true);
```

Comment:
I assume is unlikely to throw exceptions? See special behaviors lost when moving from `TailCall`. Maybe better to enhance `TailCall` to make `onFailure` non`final` so you can call `super` but still do some special processing with the exception.

Reply:
Done. I wasn't sure whether to put the call to `super.onFailure()` before or after the new logic, but I put it before the new logic and the tests still passed.